crates/bevy_ecs/src/query/state.rs

```diff
@@ -40,6 +40,27 @@ impl<Q: WorldQuery, F: WorldQuery> FromWorld for QueryState<Q, F> {
     }
 }
 
+impl<Q: WorldQuery, F: WorldQuery> QueryState<Q, F> {
+    pub(crate) fn as_immut(&self) -> &QueryState<Q::ReadOnly, F::ReadOnly> {
+        // Safety: invariant on `WorldQuery` trait upholds that `Q::ReadOnly` and `F::ReadOnly`
+        // have a subset of the access, and match the exact same archetypes/tables as `Q`/`F` respectively.
+        unsafe { self.as_whatever::<Q::ReadOnly, F::ReadOnly>() }
+    }
+
+    /// # Safety
+    ///
+    /// `NewQ` must have a subset of the access that `Q` does and match the exact same archetypes/tables
+    /// `NewF` must have a subset of the access that `F` does and match the exact same archetypes/tables
+    pub(crate) unsafe fn as_whatever<
+        NewQ: WorldQuery<State = Q::State>,
+        NewF: WorldQuery<State = F::State>,
+    >(
+        &self,
+    ) -> &QueryState<NewQ, NewF> {
+        &*(self as *const QueryState<Q, F> as *const QueryState<NewQ, NewF>)
```

Comment:
This does not spark joy. 

I am however reasonably convinced that this is sound (I haven't dared look at the mess of query traits otherwise, but I don't think this API can ever have language level UB).

I feel like it should be possible to seperate the implementation of `WorldQuery` from the state in `WorldQueryState`, somehow.

At the very least, we could split out all the non-Q/F state fields into a sub-struct, so that they are not forced into the pessimisation of `repr(C)`.

Reply:
oh splitting into a separate struct is really smart thanks

Reply:
I don't know what you mean by `WorldQueryState` theres no item named that anywhere in bevy_ecs.

Reply:
My apologies. I meant `QueryState`.